### PR TITLE
fern: SWA flat-LR tail avg — post-training weight averaging in basin

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

Stochastic Weight Averaging (SWA, Izmailov et al. 2018) applied to a **flat-LR tail** of the training run outperforms single-checkpoint EMA and post-hoc EMA-soup at short budgets. The key difference from PR #491 (EMA soup, closed null) is the training schedule:

- EMA soup (PR #491) averaged EMA snapshots from a *cosine-decaying* run — the snapshots came from very different parts of the loss landscape (epoch 1 warmup, ep2, ep3-partial), and uniform averaging produced a non-functional parameter mean.
- SWA-flat: **freeze the LR at a low constant value for the last N steps**, then uniform-average raw (non-EMA) model weights over those N steps. This is the canonical SWA setup: snapshots are taken during a flat-LR quasi-stationary phase, so they sample the same basin and averaging cuts to the geometric centre.

Literature evidence: SWA consistently beats single-checkpoint by 0.3–1.5pp on vision tasks with a 10–20% tail of the budget dedicated to flat LR. CFD surrogates have smooth loss landscapes well-suited to weight-space interpolation.

**Implementation plan:**
1. Train with cosine schedule, but instead of cosine down to zero: cosine from `lr` to `swa_lr` over the first `(1 - swa_fraction)` of budget, then hold `swa_lr` constant for the last `swa_fraction`.
2. During the flat phase, save raw model weights (not EMA shadow) every `swa_save_interval_steps` steps.
3. After timeout, average all saved snapshots parameter-wise in float32, evaluate soup on val+test.

You can reuse the `--soup-snapshots N` infra you built in PR #491 — just add:
- `--swa-lr` (default 0.0 = disabled, i.e. full cosine): set to e.g. `5e-6`
- `--swa-fraction` (default 0.2): fraction of total steps to hold flat LR
- `--swa-save-interval-steps` (default 500): how often to save raw-model snapshots during flat phase

During flat phase: replace `scheduler.step()` with a no-op; keep `optimizer.param_groups[0]["lr"] = swa_lr` constant. Save raw `model.state_dict()` (not EMA shadow) to `outputs/drivaerml/swa_snapshots/swa_step_{global_step:08d}.pt`, rotating to keep last N.

After training completes, add `--swa-eval` flag analogous to `--soup-eval`, which loads all SWA snapshots, averages them in float32, and evaluates on val+test under `swa_val/` and `swa_test/` W&B keys.

## Baseline

Current yi SOTA: **val_abupt 9.039%** (PR #309 thorfinn, `--clip-grad-norm 0.5`). Aspirational: 7.546% (PR #311, STRING-sep PE on tay).

Active merge bar: val_abupt < 9.039%.

| Metric | Yi baseline (PR #309) | AB-UPT target |
|---|---:|---:|
| val_abupt | **9.039%** | — |
| surface_pressure (test) | ~5.5% | 3.82% |
| wall_shear (test) | ~10.5% | 7.29% |
| volume_pressure (test) | ~13.5% | 6.08% |

### Step 1 — Add SWA schedule support

In `train.py`, add three new config fields:
```python
swa_lr: float = 0.0          # disabled if 0.0
swa_fraction: float = 0.20   # fraction of total steps with flat LR
swa_save_interval_steps: int = 500  # save raw-model snapshot every N flat steps
```

And CLI args:
```
--swa-lr SWA_LR
--swa-fraction SWA_FRACTION  
--swa-save-interval-steps SWA_SAVE_INTERVAL_STEPS
```

Modify the per-step LR update to detect the flat phase:
```python
if config.swa_lr > 0:
    total_steps = steps_per_epoch * max_epochs
    flat_start_step = int(total_steps * (1 - config.swa_fraction))
    if global_step >= flat_start_step:
        # Flat phase: override LR, save snapshot
        for pg in optimizer.param_groups:
            pg["lr"] = config.swa_lr
        if rank == 0 and (global_step - flat_start_step) % config.swa_save_interval_steps == 0:
            snap_dir = Path(config.output_dir) / "swa_snapshots"
            snap_dir.mkdir(parents=True, exist_ok=True)
            snap_path = snap_dir / f"swa_step_{global_step:08d}.pt"
            tmp_path = snap_path.with_suffix(".tmp")
            torch.save({"state_dict": model.state_dict()}, tmp_path)
            tmp_path.rename(snap_path)
        dist.barrier()
    else:
        scheduler.step()  # normal cosine step
else:
    scheduler.step()  # no SWA, normal cosine
```

Note: skip `scheduler.step()` during the flat phase entirely — don't call it, don't step it.

### Step 2 — Add `--swa-eval` mode

Analogous to `--soup-eval` (PR #491). When `--swa-eval` is set:
1. Load all `swa_step_*.pt` in `outputs/drivaerml/swa_snapshots/`
2. Average state_dicts param-wise in float32
3. Load into eval_model
4. Run val+test evaluation, log to `swa_val/` and `swa_test/`

This is identical to your `--soup-eval` implementation but reading from `swa_snapshots/` and loading `state_dict` (not `ema_state_dict`).

### Step 3 — Training run

Arm A (primary): SWA with `swa_fraction=0.20`, `swa_lr=5e-6`

```bash
torchrun --standalone --nproc_per_node=4 train.py \
  --agent fern \
  --wandb-group fern-r32-swa-flat-lr \
  --wandb-name "fern-r32-arm-a-swa20-5e6" \
  --optimizer lion \
  --lr 1e-4 --weight-decay 1e-4 \
  --no-compile-model --batch-size 4 \
  --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --model-slices 128 \
  --ema-decay 0.999 --lr-warmup-epochs 1 --clip-grad-norm 0.5 \
  --epochs 50 \
  --swa-lr 5e-6 --swa-fraction 0.20 --swa-save-interval-steps 500
```

After training completes and harness runs its standard EMA eval:
```bash
CUDA_VISIBLE_DEVICES=0 python train.py --swa-eval \
  --agent fern \
  --wandb-group fern-r32-swa-flat-lr \
  --wandb-name "fern-r32-arm-a-swa-eval" \
  --no-compile-model --batch-size 4 \
  --eval-surface-points 65536 --eval-volume-points 65536 \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --model-slices 128
```

### Step 4 — Reporting

Please report a four-row comparison table:

| Metric | Best EMA (single ckpt) | SWA avg (20% tail) | Δ (SWA − EMA) |
|---|---|---|---|
| val_abupt | | | |
| test_abupt | | | |
| surface_p (test) | | | |
| wall_shear (test) | | | |
| ws_y (test) | | | |
| ws_z (test) | | | |
| vol_p (test) | | | |

Plus: number of SWA snapshots averaged, steps in flat phase, W&B run IDs.

## What success looks like

- SWA avg beats single-ckpt EMA by ≥0.2pp on val_abupt → strong flat-basin signal, MERGE if also beats 9.039% yi bar
- val→test gap with SWA ≤ val→test gap with single EMA → generalization improvement
- SWA ≈ EMA → model is still far from a flat basin at this budget; reportable null, informative about convergence state

## Notes

- **SWA saves raw model weights, NOT EMA shadow** — this is the key structural difference from PR #491. EMA shadow ≈ running average of raw weights with exponential decay. SWA takes a uniform average of raw checkpoints. At flat LR, raw weights orbit the basin centre, and uniform averaging cuts to the centre of the orbit rather than tracking it exponentially.
- **swa_fraction=0.20** means the last 20% of the total step budget is flat. At ~5,441 steps/epoch × 2.3 epochs effective = ~12,500 total steps, flat starts at step ~10,000. With save_interval=500, you'll get ~5 snapshots. Adjust save_interval to get at least 4 snapshots.
- **swa_lr=5e-6** is 5% of the Lion lr=1e-4. This is a reasonable quiescent LR — enough to continue learning slightly but not enough to cause significant displacement from the basin.
- Clear the swa_snapshots directory at run start (same pattern as soup_snapshots in your PR #491 implementation).

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR: approved as-is. SWA flat-LR tail is the structurally correct follow-up to your null EMA-soup PR #491 — flat-LR snapshots sample the same basin, so uniform averaging cuts to its centre rather than averaging across loss-landscape regions.

**Updated baseline you should beat:** the active merge bar has moved. True yi SOTA is now **val_abupt 7.1792%** (PR #489 thorfinn, vol-points curriculum, run `r5rw40rn`), not 9.039%. The 9.039% figure in your PR body is stale (PR #309). Adjust your 'what success looks like' table accordingly — SWA needs to clear 7.1792% to merge. Track 'best EMA single-ckpt during your run' as your primary internal control.

**One implementation note:** save snapshots from rank 0 only (you have it right). Confirm `dist.barrier()` after the save block in all ranks (you have it). One more — clear `outputs/drivaerml/swa_snapshots/` at run start (you also note this; just calling it out).

**On budget:** at ~5,441 steps/epoch and SENPAI_MAX_EPOCHS in the 6h pod budget, you'll get ~2.4 epochs total — flat phase = 0.48 epochs ≈ 2,610 steps. With save_interval=500 that's 5–6 snapshots, sufficient for a uniform mean. If you want more density, drop save_interval to 250 (10–12 snapshots).

**Reporting nit:** add a row for `abupt_axis_mean` (val and test) to your comparison table — that's the single number the advisor evaluates against the merge bar.

Ready when you are. Mark ready-for-review when both EMA and SWA evals are logged.

**Comment by morganmcg1** [comment]
ADVISOR CORRECTION: my previous comment cited val_abupt 7.1792% from PR #489 as the active merge bar. That was an error — **PR #489 (thorfinn, vol-points curriculum) was merged into the `tay` branch, not `yi`**. The active yi merge bar is unchanged: **val_abupt 9.039%** (PR #309 thorfinn grad-clip=0.5).

7.1792% (PR #489) and 7.546% (PR #311) are aspirational tay-branch numbers, not currently reproducible on the yi codebase. Beat 9.039% to merge. Apologies for the confusion.

**Comment by morganmcg1** [comment]
STUDENT fern:

## Results

**Status: Null — SWA flat-LR tail loses by ~0.2pp on every metric vs single-checkpoint EMA at this convergence state.**

### Run details

- **Train W&B run:** `i8xi0z0g` (group `fern-r32-swa-flat-lr`)
- **SWA eval W&B run:** `tcpcrbb4` (group `fern-r32-swa-flat-lr`)
- **Total train wall time:** 275.6 min (timeout-limited; train_timeout_minutes=270, val_budget_minutes=90 of the 360-min pod budget).
- **Best EMA epoch:** 3 (final reachable; run timed out mid-epoch 3 at step 12213, so epoch 3 was the last completed validation event).
- **Peak GPU memory:** 67.5 GB / 96 GB.
- **Hardware:** torchrun `--nproc_per_node=4` (DDP world_size=4, effective batch 16).
- **Params:** 12.70M (4L/512d, 8h, 128 slices).

### SWA flat-LR phase

- `--swa-lr 5e-6 --swa-fraction 0.20 --swa-save-interval-steps 500` (advisor-approved as-is).
- Initial flat_start_step=217,680 was correctly refined post-warmup to ≈9,732 once the per-step time was measured against the 270-min train budget.
- Flat phase ran from step 9,732 → 12,213 (≈2,481 steps, ~0.46 epoch).
- **5 SWA snapshots collected** at steps 9732 / 10232 / 10732 / 11232 / 11732, each `model.state_dict()` (raw weights, NOT EMA shadow).

### SWA diagnostics

| Quantity | Value |
|---|---:|
| n_snapshots | 5 |
| snap_norm_mean | 218.28 |
| avg_norm | 218.26 |
| consecutive_l2_mean | 2.500 |
| consecutive_l2_max | 3.178 |
| snap_to_avg_l2_mean | 2.850 |
| snap_to_avg_l2_max | 4.589 |

`avg_norm ≈ snap_norm_mean` to 4 sig figs → uniform averaging did **not** collapse the parameters (good; means snapshots are clustered, not orthogonal directions). `snap_to_avg_l2_mean / snap_norm_mean ≈ 1.3 %` → snapshots orbit the mean tightly. Diagnostics look like a clean SWA-style mean; the issue is the basin, not the averaging.

### Comparison: single-ckpt EMA vs SWA-avg (20% tail, 5 snapshots)

| Metric | Best EMA (single ckpt) | SWA avg (20% tail) | Δ (SWA − EMA) |
|---|---:|---:|---:|
| **abupt_axis_mean (val)** | **11.4547** | **11.6680** | **+0.2133** |
| **abupt_axis_mean (test)** | **12.5241** | **12.7232** | **+0.1992** |
| surface_pressure (val) | 7.3502 | 7.4898 | +0.1395 |
| surface_pressure (test) | 7.0890 | 7.2328 | +0.1438 |
| wall_shear (val) | 12.8970 | 13.1095 | +0.2125 |
| wall_shear (test) | 12.7401 | 12.9573 | +0.2172 |
| ws_x (test) | 10.8720 | 11.0580 | +0.1860 |
| ws_y (test) | 15.6923 | 15.9524 | +0.2600 |
| ws_z (test) | 15.4627 | 15.7352 | +0.2725 |
| volume_pressure (val) | 6.9423 | 7.1271 | +0.1848 |
| volume_pressure (test) | 13.5043 | 13.6378 | +0.1335 |

All metrics worse by 0.13–0.31pp under SWA. The val→test gap is essentially identical (EMA: 11.45→12.52 = +1.07pp; SWA: 11.67→12.72 = +1.06pp), so SWA neither helps nor hurts generalization — it just shifts both numbers up uniformly.

### What happened

This is the "SWA ≈ EMA → still far from a flat basin" branch of the success matrix the advisor outlined, except SWA is in fact slightly *behind* EMA. The model is at val_abupt 11.5% in this run, whereas the yi merge bar (PR #309) is val_abupt 9.039% reached after **9 full epochs ≈ 23,544 steps** on 8 GPUs. Our run only completed ~2.25 epochs / 12,213 steps before the train budget ran out, so the optimization is still well inside the steep-descent regime — at epoch 3 the val_abupt was still falling 8pp/epoch (epoch 1: 23.4 → epoch 3: 11.5). Snapshots taken inside that descent window orbit a moving target, not a basin centre, and uniform averaging in moving-target conditions is exactly what hurt EMA-soup in PR #491.

The SWA diagnostics support this: consecutive_l2 values (2.5–3.2) are large enough that the model is still drifting between save points despite the 50× LR drop. Either the basin is genuinely not flat at this point in training, or `swa_lr=5e-6` is still high enough that Lion is pushing the weights along a local descent direction rather than orbiting a minimum. EMA's exponential decay (0.999) is implicitly a tighter average over recent steps when the weights are still drifting, so it tracks the descent better than a uniform mean over a 2.5k-step window.

**Neither EMA nor SWA beats the merge bar (9.039%).** The fundamental issue is that this run was budget-limited to 3 epochs. SWA could only ever help on top of an already-converged training run — it cannot rescue an under-trained one.

### Suggested follow-ups

1. **Re-run SWA on a longer run** that actually reaches sub-9.039% val_abupt before the flat phase begins. PR #222's data shows val curves still steep at step 12k (11.6% there → 9.3% at step 23.5k), so extending wall-clock budget (or sub-sampling to land 8 GPUs effectively) would put the flat phase inside a genuinely shallow region. SWA is most likely to win on a long Round-18 STRING-sep run on tay/yi where val_abupt has plateaued.
2. **Try `swa_lr=1e-6`** (10× drop). At 5e-6 the diagnostics show ~1% norm drift between consecutive snapshots — too noisy. A flat LR closer to "stop learning, just sample" might let the snapshots actually orbit.
3. **Average EMA shadow snapshots, not raw weights** during flat phase. EMA shadow is already a smoothed estimate; uniform-averaging EMA shadow over the flat tail is essentially "nested smoothing" and should be more conservative than averaging raw drifting weights. This would also mean the SWA result inherits EMA's already-better baseline.
4. **Smaller `swa_fraction` (e.g. 0.05–0.10)** — since the model is still descending, take a tighter average just at the very end where val_abupt has flattened most.

### Train command (Arm A)

```bash
torchrun --standalone --nproc_per_node=4 train.py \
  --agent fern \
  --wandb-group fern-r32-swa-flat-lr \
  --wandb-name "fern-r32-arm-a-swa20-5e6" \
  --optimizer lion \
  --lr 1e-4 --weight-decay 1e-4 \
  --no-compile-model --batch-size 4 \
  --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --model-slices 128 \
  --ema-decay 0.999 --lr-warmup-epochs 1 --clip-grad-norm 0.5 \
  --epochs 50 \
  --swa-lr 5e-6 --swa-fraction 0.20 --swa-save-interval-steps 500
```

### SWA-eval command

```bash
CUDA_VISIBLE_DEVICES=0 python train.py --swa-eval \
  --agent fern \
  --wandb-group fern-r32-swa-flat-lr \
  --wandb-name "fern-r32-arm-a-swa-eval" \
  --no-compile-model --batch-size 4 \
  --eval-surface-points 65536 --eval-volume-points 65536 \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --model-slices 128
```

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #527 because ADVISOR: Closing as informative null. The SWA diagnostics were impeccable — 5 clean snapshots, avg_norm≈snap_norm_mean to 4 sig figs, tight orbit (snap_to_avg_l2 ~1.3% of norm). The mechanism was correctly implemented. The failure mode is not the averaging — it is that SWA requires a *converged* model in a flat basin, and this run only completed 2.25 epochs (val_abupt 11.45% EMA) vs the yi merge bar of 9.039%, which requires ~9 epochs (23,544 steps on 8 GPUs). The model was still in the steep-descent regime when the flat phase began; consecutive_l2 ~2.5–3.2 confirms it was still drifting, not orbiting. Fern's analysis is exactly right. SWA is not ruled out in principle — it is ruled out as a short-budget trick. If we ever run a full convergence run that reaches <9% before the flat phase, SWA could compound a further 0.2–0.5pp. For now the higher-ROI path is the STRING-sep PE port (PR #490), which directly addresses the architectural deficit between yi and tay (7.546% vs 9.039%)..
